src/profile/references/ref_macro.rs

```diff
@@ -0,0 +1,173 @@
+/// Declares a `uniffi::Object` exported type holding
+/// an inner type with an `RwLock` so that we can
+/// take the value from an `Arc<Ref>`, thus avoiding
+/// expensive clone operations.
+macro_rules! decl_ref_named {
+    (
+        $struct_name: ty,
+        $inner: ty
+    ) => {
+        paste! {
+            use crate::prelude::*;
+            use std::sync::{Arc, RwLock};
+
+            #[derive(Debug, uniffi::Object)]
+            #[uniffi::export(Debug, Eq, Hash)]
+            pub(crate) struct $struct_name {
+                inner: RwLock<Option<$inner>>,
+            }
+
+            impl $struct_name {
+                pub(crate) fn with_inner(inner: $inner) -> Self {
+                    Self {
+                        inner: RwLock::new(Some(inner)),
+                    }
+                }
+            }
+
+            impl std::hash::Hash for $struct_name {
+                fn hash<H>(&self, state: &mut H)
+                where
+                    H: std::hash::Hasher,
+                {
+                    match self.inner.read() {
+                        Ok(ref guard) => {
+                            state.write_u8(1);
+                            match guard.as_ref() {
+                                Some(prof) => {
+                                    prof.hash(state);
+                                    state.write_u8(100);
+                                }
+                                None => state.write_u8(200),
+                            }
+                        }
+                        _ => {
+                            state.write_u8(255);
+                        }
+                    }
+                }
+            }
+
+            impl Eq for $struct_name {}
+            impl PartialEq for $struct_name {
+                fn eq(&self, other: &Self) -> bool {
+                    match (self.inner.try_read(), other.inner.try_read()) {
+                        (Ok(ref lhs), Ok(ref rhs)) => match(lhs.as_ref(), rhs.as_ref()) {
+                            (Some(l), Some(r)) => l == r,
+                            _ => false
+                        },
+                        _ => false,
+                    }
+                }
+            }
+
+            impl From<$inner> for $struct_name {
+                fn from(value: $inner) -> Self {
+                    Self::with_inner(value)
+                }
+            }
+
+            #[uniffi::export]
+            impl $struct_name {
+                #[uniffi::constructor]
+                pub(crate) fn new(inner: $inner) -> Arc<Self> {
+                    Arc::new(Self::from(inner))
+                }
+
+                pub(crate) fn take(self: Arc<Self>) -> Result<$inner> {
+                    self.inner
+                        .try_write()
+                        .expect("Should only acquire write lock once.")
+                        .take()
```

Comment:
the `take` call on the RwLockWriteGuard here replaces `Some` with `None` and returns the value. This consuming this reference type and returning the inner value. 

This function cannot be called twice in Rust, thanks to `self` being consumed (not passed by reference) - so we are protected in the land of Rust. And it returns `Result` so is throwing in the FFI land (and WILL throw if called twice in FFI land).